lib/Modules/ComputeHooks/ComputeHooks.psm1

```diff
@@ -856,3 +858,131 @@ function Invoke-LocalMonitorsRelationJoined {
         Set-JujuRelation -RelationId $rid -Settings $settings
     }
 }
+
+function Invoke-HGSRelationJoined {
+    $adCtxt = Get-ActiveDirectoryContext
+    if(!$adCtxt.Count) {
+        Write-JujuWarning "AD context is not ready yet"
+        return
+    }
+
+    $domainUser = "{0}\{1}" -f @($adCtxt['domainName'], $adCtxt['username'])
+    $securePass = ConvertTo-SecureString $adCtxt['password'] -AsPlainText -Force
+    $adCredential = New-Object System.Management.Automation.PSCredential($domainUser, $securePass)
+    $session = New-CimSession -Credential $adCredential
+
+    $adGroupName = Get-JujuCharmConfig -Scope 'ad-computer-group'
+    $adGroup = Get-CimInstance -ClassName "Win32_Group" -Filter "Name='$adGroupName'" -CimSession $session
+    $relationSettings = @{
+        'ad-address' = $adCtxt['address']
+        'ad-domain-name' = $adCtxt['domainName']
+        'ad-user'= $adCtxt['username']
+        'ad-user-password' = $adCtxt['password']
+        'ad-group-sid' = $adGroup.SID
+    }
+
+    $rids = Get-JujuRelationIds -Relation 'hgs'
+    foreach($rid in $rids) {
+        Set-JujuRelation -RelationId $rid -Settings $relationSettings
+    }
+}
+
+function Invoke-HGSRelationChanged {
+    $ctxt = Get-HGSContext
+    if(!$ctxt.Count) {
+        Write-JujuWarning "HGS context is not ready yet"
+        return
+    }
+
+    Write-JujuWarning "Installing required HGS features"
+    Install-WindowsFeatures -Features @('HostGuardian', 'RSAT-Shielded-VM-Tools', 'FabricShieldedTools')
+
+    $nameservers = Get-CharmState -Namespace "novahyperv" -Key "nameservers"
+    if(!$nameservers) {
+        # Save the current DNS nameservers before pointing the DNS to the HGS server
+        $nameservers = Get-PrimaryAdapterDNSServers
+        Set-CharmState -Namespace "novahyperv" -Key "nameservers" -Value $nameservers
+    }
+
+    Set-DnsClientServerAddress -InterfaceAlias (Get-MainNetadapter) -Addresses @($ctxt['hgs-private-ip'])
+
+    $domain = $ctxt['hgs-domain-name']
+    Set-HgsClientConfiguration -AttestationServerUrl "http://$domain/Attestation" `
+                               -KeyProtectionServerUrl "http://$domain/KeyProtection" -Confirm:$false
+}
+
+function Invoke-HGSRelationDeparted {
+    # Restore the DNS'es saved before pointing the DNS to the HGS server
+    $nameservers = Get-CharmState -Namespace "novahyperv" -Key "nameservers"
+    if($nameservers) {
+        Set-DnsClientServerAddress -InterfaceAlias (Get-MainNetadapter) -Addresses $nameservers
+        Remove-CharmState -Namespace "novahyperv" -Key "nameservers"
+    }
+}
+
+function Invoke-AMQPRelationJoinedHook {
+    $username, $vhost = Get-RabbitMQConfig
+
+    $relationSettings = @{
+        'username' = $username
+        'vhost' = $vhost
+    }
+
+    $rids = Get-JujuRelationIds -Relation "amqp"
+    foreach ($rid in $rids){
+        Set-JujuRelation -RelationId $rid -Settings $relationSettings
+    }
+}
+
+function Invoke-MySQLDBRelationJoinedHook {
```

Comment:
Sorry for late reply. This used to be in `OpenStackCommon`, but it's just a main function to be used in the hooks. As per last syncs, we discussed that common libraries mustn't have any main functions. Thus I moved it to the main functions charm module.